### ISSUE TITLE
Possibly better names for the `as_*` functions

### ISSUE THREAD

**Issue by barosl** [serde-rs/json#39]
It's probably just me, but I've *always* been confused when trying to recall the names of the `as_*` functions on `Value`.

I first try to use `.as_str`, because that name is popular in `std` for methods returning a `&str`. Then I fail, and next I try `.as_string` and it succeeds. The problem is, after that I try to call `.as_integer` because that was the logic I used in the previous stage ("use the fully qualified name"). Then I fail again, so instead I try `.as_number` because that's the terminology used in the JSON spec. And I fail once more. At this point I give up and open the browser to read the documentation, realizing the method was `.as_i64`.

It seems that those names were inherited from `rustc-serialize`. I can see the rationale used to decide the current names, which corresponds to the names of the enum variants (`I64`, `U64`, `F64`, `String`, `Boolean`). But personally, I think using `.as_<resulting_type>` as the names has several advantages:

- One can know intuitively what type will be returned. `.as_str` returns `&str`, `.as_i64` returns `i64`, and `.as_bool` returns `bool`, and so on.
- If we add more "helper" methods in the future, such as `.as_i32` or `.as_i16`, it would be straightforward. Even `as_str` (the one returning `&str`) and `as_string` (the one returning `String`) could coexist!
- They are shorter!

The only (but possibly big) disadvantage is, of course, it is a breaking change. (And I missed the 0.7 timeframe due to my laziness!) But this could be alleviated by introducing a deprecation period.

What's your thoughts on this?

**Comment by dtolnay** [comment]
To be specific, the required changes are:

- as_string to as_str (this matches [`String::as_str`](https://doc.rust-lang.org/std/string/struct.String.html#method.as_str))
- as_boolean to as_bool

**Comment by erickt** [comment]
@barosl / @dtolnay: The reason I named it this way was to match how the json spec defines these terms. That's why I used `Array` instead of `Vec`. However, I'm not stuck to those names, and it'd probably be better to stick with the Rust conventions, rather than json. So, @dtolnay, those two changes sound fine to me. We should also consider renaming `Value::Array` to `Value::Vec`, and `as_array` to `as_slice`.
